### PR TITLE
some TODOs in Iterator fixed #487

### LINKED ISSUE #487: Make javaslang.collection.Iterator a Traversable
To do this, we (re-)add TraversableOnce. All methods of Traversable are move to TraversableOnce and Traversable extends TraversableOnce.

Then Iterator extends TraversableOnce.

Additionally we add two methods to Traversable~~Once~~:

* `boolean hasDefiniteSize()` // ~~Iterator,~~ Stream ~~, Gen and Arbitrary~~ return(s) false, all others return true
* `boolean isTraversableAgain()` // ~~Iterator returns false,~~ all (others) return true

Note: TraversableOnce has undefined size and is not traversable again by definition.

### PR DESCRIPTION
**Pull request by ruslansennov**

### CODE REVIEW

File: src/main/java/javaslang/collection/HashSet.java [hunk 1]
```diff
@@ -186,7 +186,7 @@ public boolean contains(T element) {
 
     @Override
     public Option<T> findLast(Predicate<? super T> predicate) {
-        throw new UnsupportedOperationException("TODO");
+        return findFirst(predicate);
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/HashSet.java]: great!!

File: src/main/java/javaslang/collection/Iterator.java [hunk 2]
```diff
@@ -1334,8 +1452,35 @@ public T next() {
...
+            return empty();
+        } else {
+            final Iterator<T> that = this;
+            return new Iterator<T>() {
+                private Queue<T> queue = Queue.empty();
+
+                @Override
+                public boolean hasNext() {
+                    while (that.hasNext()) {
+                        queue = queue.append(that.next());
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Iterator.java]: I would call `queue.enqueue(that.next())` (which is the same)

**Comment by danieldietrich** [comment]
Thx for taking care of HashSet & Iterator!

Today we celebrated the birthday of my 5 yo daugther - 12 kids(!) where there. Now I'm relaxing by implementing the RedBlackTree :)

**Comment by ruslansennov** [comment]
> 12 kids(!) where there

local hell

**Comment by danieldietrich** [comment]
yep

**Pull request merged**